### DISCUSSION TITLE
How to upload a file using GitHub API

### DISCUSSION THREAD

**Discussion by jkalandaibm** [community/community#24723] [Apps, API and Webhooks]
<p>Maybe it’s just me, but this guide <a href="https://developer.github.com/v3/repos/contents/" rel="nofollow noopener">https://developer.github.com/v3/repos/contents/</a> is incredibly vague.<br>
Simple examples would be handy, and not so much emphasis on what the return messages.</p>
<p>The objective is to upload a file, preferably .pdf or .png through a curl command.<br>
Any help/pointers will be highly appreciated.</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>It’s hard to write documentation that is both approachable but also has a good amount of detail to answer whatever questions that people will have.</p>
<p>So you’re going to need to understand a few things about the information in the <a href="https://developer.github.com/v3/repos/contents/#create-a-file">create a file endpoint</a> reference:</p>
<ul>
<li>You’re going to need to be authenticated and have write permissions on the repository</li>
<li>You need to pass certain things in the URL and others in the body of the HTTP request
<ul>
<li>Replaceable parameters in the URL start with a colon <code>:</code>
</li>
</ul>
</li>
</ul>
<p>I’m also going to be using <a href="https://httpie.org">HTTPie</a> instead of curl because it makes JSON REST API interactions much easier. (And I was never very good at using curl for things :grinning:)</p>
<p>I used the following command:</p>
<pre><code class="lang-shell">http PUT https://api.github.com/repos/lee-dohm/test-repo/contents/hello.txt \
  "Authorization: token REDACTED-TOKEN" \
  message="my commit message" \
  committer:="{ \"name\": \"Lee Dohm\", \"email\": \"1038121+lee-dohm@users.noreply.github.com\" }" \
  content="bXkgbmV3IGZpbGUgY29udGVudHM="
</code></pre>
<p>which resulted in the response:</p>
<pre><code class="lang-json">{
  "commit": {
    "author": {
      "date": "2019-05-28T23:10:17Z",
      "email": "1038121+lee-dohm@users.noreply.github.com",
      "name": "Lee Dohm"
    },
    "committer": {
      "date": "2019-05-28T23:10:17Z",
      "email": "1038121+lee-dohm@users.noreply.github.com",
      "name": "Lee Dohm"
    },
    "html_url": "https://github.com/lee-dohm/test-repo/commit/76c36cec967c4d647422cf41982829d153fa8cce",
    "message": "my commit message",
    "node_id": "MDY6Q29tbWl0MTEyMTM3MDU2Ojc2YzM2Y2VjOTY3YzRkNjQ3NDIyY2Y0MTk4MjgyOWQxNTNmYThjY2U=",
    "parents": [
      {
        "html_url": "https://github.com/lee-dohm/test-repo/commit/d775caeeac9a48eb73f8012349c36a8c2b8c6e63",
        "sha": "d775caeeac9a48eb73f8012349c36a8c2b8c6e63",
        "url": "https://api.github.com/repos/lee-dohm/test-repo/git/commits/d775caeeac9a48eb73f8012349c36a8c2b8c6e63"
      }
    ],
    "sha": "76c36cec967c4d647422cf41982829d153fa8cce",
    "tree": {
      "sha": "1f3c23004c5252ee11cb723e3319d26d91bc19ac",
      "url": "https://api.github.com/repos/lee-dohm/test-repo/git/trees/1f3c23004c5252ee11cb723e3319d26d91bc19ac"
    },
    "url": "https://api.github.com/repos/lee-dohm/test-repo/git/commits/76c36cec967c4d647422cf41982829d153fa8cce",
    "verification": {
      "payload": null,
      "reason": "unsigned",
      "signature": null,
      "verified": false
    }
  },
  "content": {
    "_links": {
      "git": "https://api.github.com/repos/lee-dohm/test-repo/git/blobs/0d5a690c8fad5e605a6e8766295d9d459d65de42",
      "html": "https://github.com/lee-dohm/test-repo/blob/master/hello.txt",
      "self": "https://api.github.com/repos/lee-dohm/test-repo/contents/hello.txt?ref=master"
    },
    "download_url": "https://raw.githubusercontent.com/lee-dohm/test-repo/master/hello.txt",
    "git_url": "https://api.github.com/repos/lee-dohm/test-repo/git/blobs/0d5a690c8fad5e605a6e8766295d9d459d65de42",
    "html_url": "https://github.com/lee-dohm/test-repo/blob/master/hello.txt",
    "name": "hello.txt",
    "path": "hello.txt",
    "sha": "0d5a690c8fad5e605a6e8766295d9d459d65de42",
    "size": 20,
    "type": "file",
    "url": "https://api.github.com/repos/lee-dohm/test-repo/contents/hello.txt?ref=master"
  }
}
</code></pre>
<p>and generated <a href="https://github.com/lee-dohm/test-repo/commit/76c36cec967c4d647422cf41982829d153fa8cce">this commit</a>.</p>
<p>I hope that helps!</p>

**Reply by SH20RAJ** [reply]
How to upload to github releases using api php/nodejs

**Comment by kovalensky** [comment]
<p>Maybe it will help someone:</p>
<pre><code>&lt;?php
$file_git = "wall.jpg";
$data_git = array(
'sha'=&gt;file_get_contents("sha.txt"),
'message'=&gt;'image',
'content'=&gt; base64_encode($file_git),
'committer'=&gt; array(
'name'=&gt;'Jacob',
'email' =&gt; '45331093+greenandgreen@users.noreply.github.com'
)
);
$data_string_git = json_encode($data_git);
$ch_git = curl_init('https://api.github.com/repos/YOUR_REPO/contents/wall.jpg');
curl_setopt($ch_git, CURLOPT_CUSTOMREQUEST, "PUT");
curl_setopt($ch_git, CURLOPT_POSTFIELDS, $data_string_git);
curl_setopt($ch_git, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch_git, CURLOPT_HTTPHEADER, array(
'Content-Type: application/json',
'User-Agent: Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 YaBrowser/19.9.3.314 Yowser/2.5 Safari/537.36',
'Authorization: token PLACE_YOUR_PERSONAL_TOKEN_HERE'
));
$result_git = curl_exec($ch_git);
echo $result_git;
$p_git = json_decode($result_git);
file_put_contents("sha.txt",$p_git-&gt;content-&gt;sha);
?&gt;
</code></pre>
<p>Tested on PHP 7.<br>
If the code is working, then I wouldn’t refuse likes, in gratitude for the little work :slight_smile:</p>

**Reply by rushikesh456thorat** [reply]
bro can you give ans for Javascript

**Comment by qaiserpc** [comment]
<p>what is </p>
<pre><code>'sha'=&gt;file_get_contents("sha.txt")
</code></pre>
<p>To upload a new file do you still need it? My problem is that with HTTPFul i manage to put the zip file into github repo but the the file is corrupt, when i I do git pull and try to unzip the file it say unsupported format. and the size of the file is also just 4K. Here is my code </p>
<pre><code>$response = \Httpful\Request::put($url)  
 -\&gt;addHeader('Authorization', "Basic $base64")  
 -\&gt;addHeaders(  
array(  
'Accept-Encoding' =\&gt; 'gzip, deflate',  
'Host' =\&gt; 'api.github.com',  
'Cache-Control' =\&gt; 'no-cache',  
'Accept' =\&gt; '\*/\*',  
 )  
 )  
 -\&gt;body('{  
 "message": "'.$commitMessage.'",  
 "content": "'.$sourceFile.'",  
 "branch":"'.$this-\&gt;getSettings()-\&gt;branch.'"  
 }')  
 -\&gt; send();
</code></pre>
<p>Any help/suggestion will be appreciated thanks</p>

**Comment by NamanZelawat** [comment]
<p>Can anyone please show this using postman or Node.js .</p>

**Comment by joshdangdesigns** [comment]
<p>I’m trying to use the script above this comment but felt like commenting the solution to this problem since it was never listed.  Basically, you have to get the sha from your existing file.  In PHP, this is done using the sha1_file() method.</p>
<pre><code class="lang-auto">$shafile = "sha.txt";

// Put sha code into the file and close it.
$shatxtfile = fopen($shafile, "w") or die("Unable to open sha file!");
$sha = sha1_file("fakefilename.txt");
fwrite($shatxtfile, $sha);
fclose($shatxtfile);
</code></pre>

**Comment by joshdangdesigns** [comment]
<p>For those looking for this solution, this is for a public folder and only sends the name of the file.</p>

**Comment by joshdangdesigns** [comment]
<p>Also, the curl_init must specify the user, then the repository and, after contents, specify the file path.</p>

**Comment by sushant-r** [comment]
<p>couldn’t able to upload images.</p>
